types.go

```diff
@@ -2125,3 +2150,41 @@ func (AccountKeyType) ID() string {
 func (t AccountKeyType) Equal(other Type) bool {
 	return t == other
 }
+
+func generateID(location common.Location, identifier string) string {
+	if location == nil {
+		return identifier
+	}
+
+	return string(location.TypeID(nil, identifier))
+}
+
+// TypeWithCachedTypeID recursively caches type ID of type t.
+// This is needed because each type ID is lazily cached on
+// its first use in ID() to avoid performance penalty.
+func TypeWithCachedTypeID(t Type) Type {
```

Comment:
_Note: This method is only used for tests, but is added here to avoid circular dependencies._